File: src/main/java/com/ggardet/patient/internal/domain/MedicalRecordServiceImpl.java
package com.ggardet.patient.internal.domain;

import com.ggardet.patient.internal.entity.MedicalRecord;
import com.ggardet.patient.internal.repository.MedicalRecordRepository;
import com.ggardet.patient.spi.MedicalRecordService;
import com.ggardet.patient.spi.dto.MedicalRecordDTO;
import com.ggardet.patient.spi.exception.MedicalRecordNotFoundException;
import com.ggardet.patient.spi.mapper.MedicalRecordMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional
public class MedicalRecordServiceImpl implements MedicalRecordService {
    private final MedicalRecordRepository medicalRecordRepository;
    private final MedicalRecordMapper medicalRecordMapper;

    @Override
    public MedicalRecordDTO create(final MedicalRecordDTO medicalRecordDTO) {
        final var medicalRecord = medicalRecordMapper.toEntity(medicalRecordDTO);
        final var savedRecord = medicalRecordRepository.save(medicalRecord);
        return medicalRecordMapper.toDto(savedRecord);
    }

    @Override
    @Transactional(readOnly = true)
    public MedicalRecordDTO findById(final UUID id) {
        final var medicalRecord = medicalRecordRepository.findById(id)
                .orElseThrow(() -> new MedicalRecordNotFoundException(id));
        return medicalRecordMapper.toDto(medicalRecord);
    }

    @Override
    @Transactional(readOnly = true)
    public List<MedicalRecordDTO> findAll() {
        final var records = medicalRecordRepository.findAll();
        return medicalRecordMapper.toDtos(records);
    }

    @Override
    @Transactional(readOnly = true)
    public List<MedicalRecordDTO> findByPatientId(final UUID patientId) {
        final var records = medicalRecordRepository.findByPatientId(patientId);
        return medicalRecordMapper.toDtos(records);
    }

    @Override
    @Transactional(readOnly = true)
    public List<MedicalRecordDTO> findByDoctorId(final UUID doctorId) {
        final var records = medicalRecordRepository.findByDoctorId(doctorId);
        return medicalRecordMapper.toDtos(records);
    }

    @Override
    @Transactional(readOnly = true)
    public List<MedicalRecordDTO> findByDateRange(final LocalDateTime start, final LocalDateTime end) {
        final var records = medicalRecordRepository.findByRecordDateBetween(start, end);
        return medicalRecordMapper.toDtos(records);
    }

    @Override
    @Transactional(readOnly = true)
    public List<MedicalRecordDTO> findByFollowUpRequired(final Boolean followUpRequired) {
        final var records = medicalRecordRepository.findByFollowUpRequired(followUpRequired);
        return medicalRecordMapper.toDtos(records);
    }

    @Override
    public MedicalRecordDTO update(final UUID id, final MedicalRecordDTO medicalRecordDTO) {
        final var existingRecord = medicalRecordRepository.findById(id)
                .orElseThrow(() -> new MedicalRecordNotFoundException(id));
        updateRecordFields(existingRecord, medicalRecordDTO);
        final var updatedRecord = medicalRecordRepository.save(existingRecord);
        return medicalRecordMapper.toDto(updatedRecord);
    }

    @Override
    public void deleteById(final UUID id) {
        if (!medicalRecordRepository.existsById(id)) {
            throw new MedicalRecordNotFoundException(id);
        }
        medicalRecordRepository.deleteById(id);
    }

    private void updateRecordFields(final MedicalRecord existingRecord, final MedicalRecordDTO dto) {
        if (Objects.nonNull(dto.patientId())) {
            existingRecord.setPatientId(dto.patientId());
        }
        if (Objects.nonNull(dto.doctorId())) {
            existingRecord.setDoctorId(dto.doctorId());
        }
        if (Objects.nonNull(dto.recordDate())) {
            existingRecord.setRecordDate(dto.recordDate());
        }
        if (Objects.nonNull(dto.diagnosis())) {
            existingRecord.setDiagnosis(dto.diagnosis());
        }
        if (Objects.nonNull(dto.treatment())) {
            existingRecord.setTreatment(dto.treatment());
        }
        if (Objects.nonNull(dto.notes())) {
            existingRecord.setNotes(dto.notes());
        }
        if (Objects.nonNull(dto.followUpRequired())) {
            existingRecord.setFollowUpRequired(dto.followUpRequired());
        }
    }
}

File: src/main/kotlin/com/ggardet/patient/internal/domain/MedicalRecordServiceImpl.kt
package com.ggardet.patient.internal.domain

import com.ggardet.patient.internal.entity.MedicalRecord
import com.ggardet.patient.internal.repository.MedicalRecordRepository
import com.ggardet.patient.spi.MedicalRecordService
import com.ggardet.patient.spi.dto.MedicalRecordDTO
import com.ggardet.patient.spi.exception.MedicalRecordNotFoundException
import com.ggardet.patient.spi.mapper.MedicalRecordMapper
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.*

@Service
@Transactional
class MedicalRecordServiceImpl(
    private val medicalRecordRepository: MedicalRecordRepository,
    private val medicalRecordMapper: MedicalRecordMapper
) : MedicalRecordService {

    override fun create(medicalRecordDTO: MedicalRecordDTO): MedicalRecordDTO {
        val medicalRecord = medicalRecordMapper.toEntity(medicalRecordDTO)
        val savedRecord = medicalRecordRepository.save(medicalRecord)
        return medicalRecordMapper.toDto(savedRecord)
    }

    @Transactional(readOnly = true)
    override fun findById(id: UUID): MedicalRecordDTO {
        val medicalRecord = medicalRecordRepository.findById(id)
            .orElseThrow { MedicalRecordNotFoundException(id) }
        return medicalRecordMapper.toDto(medicalRecord)
    }

    @Transactional(readOnly = true)
    override fun findAll(): List<MedicalRecordDTO> {
        val records = medicalRecordRepository.findAll()
        return medicalRecordMapper.toDtos(records)
    }

    @Transactional(readOnly = true)
    override fun findByPatientId(patientId: UUID): List<MedicalRecordDTO> {
        val records = medicalRecordRepository.findByPatientId(patientId)
        return medicalRecordMapper.toDtos(records)
    }

    @Transactional(readOnly = true)
    override fun findByDoctorId(doctorId: UUID): List<MedicalRecordDTO> {
        val records = medicalRecordRepository.findByDoctorId(doctorId)
        return medicalRecordMapper.toDtos(records)
    }

    @Transactional(readOnly = true)
    override fun findByDateRange(start: LocalDateTime, end: LocalDateTime): List<MedicalRecordDTO> {
        val records = medicalRecordRepository.findByRecordDateBetween(start, end)
        return medicalRecordMapper.toDtos(records)
    }

    @Transactional(readOnly = true)
    override fun findByFollowUpRequired(followUpRequired: Boolean): List<MedicalRecordDTO> {
        val records = medicalRecordRepository.findByFollowUpRequired(followUpRequired)
        return medicalRecordMapper.toDtos(records)
    }

    override fun update(id: UUID, medicalRecordDTO: MedicalRecordDTO): MedicalRecordDTO {
        val existingRecord = medicalRecordRepository.findById(id)
            .orElseThrow { MedicalRecordNotFoundException(id) }
        updateRecordFields(existingRecord, medicalRecordDTO)
        val updatedRecord = medicalRecordRepository.save(existingRecord)
        return medicalRecordMapper.toDto(updatedRecord)
    }

    override fun deleteById(id: UUID) {
        if (!medicalRecordRepository.existsById(id)) {
            throw MedicalRecordNotFoundException(id)
        }
        medicalRecordRepository.deleteById(id)
    }

    private fun updateRecordFields(existingRecord: MedicalRecord, dto: MedicalRecordDTO) {
        existingRecord.patientId = dto.patientId
        existingRecord.doctorId = dto.doctorId
        existingRecord.recordDate = dto.recordDate
        dto.diagnosis?.let { existingRecord.diagnosis = it }
        dto.treatment?.let { existingRecord.treatment = it }
        dto.notes?.let { existingRecord.notes = it }
        dto.followUpRequired?.let { existingRecord.followUpRequired = it }
    }
}